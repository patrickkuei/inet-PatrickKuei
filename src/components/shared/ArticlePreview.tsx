import Dot from '@inet/components/shared/Dot'
import { AnonymIcon, EyeIcon, HeartIcon } from '@inet/icons'
import { IArticleViewModel } from '@inet/services/types/articles/i-article.view-model'

export interface IArticlePreviewProps {
  article: IArticleViewModel
  highlightText?: string
  createdAt: string
}

const getMatchedSubstringIndex = (
  rowString: string | undefined,
  rowSubstring: string | undefined,
) => {
  if (!rowString || !rowSubstring) {
    return -1
  }

  const [string, substring] = [
    rowString.toLowerCase(),
    rowSubstring.toLowerCase(),
  ]

  for (let i = 0; i < string.length; i++) {
    let matchedIdx = -1
    let idx = 0
    let matched = true

    if (string[i] === substring[idx]) {
      matchedIdx = i

      while (idx < substring.length - 1) {
        if (string[++i] !== substring[++idx]) {
          matched = false
          break
        }
      }

      if (matched) {
        return matchedIdx
      }

      i--
    }
  }

  return -1
}

export default function ArticlePreview({
  article,
  highlightText,
  createdAt,
}: IArticlePreviewProps) {
  const getMarkedJsx = (string: string) => {
    const matchedIdx = getMatchedSubstringIndex(string, highlightText)

    if (matchedIdx < 0) {
      return string
    } else {
      return (
        <>
          {string.slice(0, matchedIdx)}
          <span className="text-secondary-500">
            {string.slice(matchedIdx, matchedIdx + highlightText!.length)}
          </span>
          {string.slice(matchedIdx + highlightText!.length)}
        </>
      )
    }
  }

  return (
    <div className="text-left h-24 flex flex-row gap-x-2">
      <div className="w-24 shrink-0 bg-cyan-50 flex items-center justify-center">
        <img
          className="max-h-full"
          src={article.coverImageUrl}
          alt={article.title}
        />
      </div>
      <div className="flex flex-col grow gap-y-1 min-w-0">
        <div className="flex items-center gap-x-1 text-xs">
          <div className="flex items-center gap-x-1 shrink-0">
            <img
              className="w-4 h-4 border border-gray-100 rounded inline"
              src={article.category.imageUrl}
              alt={article.category.code}
            />
            <span>{article.category.code}</span>
          </div>
          <Dot />
          <div className="flex items-center gap-x-1 min-w-0">
            {article.author.avatarUrl ? (
              <img
                className="w-4 h-4 border border-gray-100 rounded-full inline"
                src={article.author.avatarUrl}
                alt={article.author.name}
              />
            ) : (
              <AnonymIcon />
            )}
            <span className="truncate">
              {getMarkedJsx(article.author.name)}
            </span>
          </div>
          <Dot />
          <span className="whitespace-nowrap">{createdAt}</span>
        </div>
        <div>
          <div className="w-fit text-tiny font-bold line-clamp-2 hover:text-primary-400 hover:cursor-pointer">
            {getMarkedJsx(article.title)}
          </div>
          <div className="text-xs text-gray-300 truncate">
            {getMarkedJsx(article.body)}
          </div>
        </div>
        <div className="text-mini flex grow items-end gap-x-4">
          <span className="flex gap-x-1">
            <EyeIcon />
            {article.views}
          </span>
          <span className="flex gap-x-1">
            <HeartIcon />
            {article.likes}
          </span>
        </div>
      </div>
    </div>
  )
}
