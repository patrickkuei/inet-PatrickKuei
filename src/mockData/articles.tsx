export type IArticle = {
  id: number
  category: {
    code: string
    imageUrl: string
  }
  author: {
    id: number
    username: string
    name: string
    avatarUrl: string
    createdAt: string
  }
  coverImageUrl: string
  title: string
  body: string
  views: number
  likes: number
  likeId: number | null
  createdAt: string
  lastModifiedAt: string | null
}

const articles: IArticle[] = [
  {
    id: 1,
    category: {
      code: 'animal',
      imageUrl: 'https://i.imgur.com/JR2xyV6s.jpg',
    },
    author: {
      id: 1,
      username: 'user0',
      name: '人類一號',
      avatarUrl: 'https://i.imgur.com/1yJEtrqs.jpeg',
      createdAt: '2022-06-27T06:32:36.929Z',
    },
    coverImageUrl: 'https://i.imgur.com/bqgW6JXh.jpeg',
    title: 'Hello world',
    body: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut a',
    views: 123,
    likes: 45,
    likeId: 1,
    createdAt: '2022-09-04T15:34:38.231Z',
    lastModifiedAt: null,
  },
  {
    id: 2,
    category: {
      code: 'sports',
      imageUrl: 'https://i.imgur.com/s4PDhtLs.jpg',
    },
    author: {
      id: 1,
      username: 'user0',
      name: '人類一號',
      avatarUrl: 'https://i.imgur.com/1yJEtrqs.jpeg',
      createdAt: '2022-06-27T06:32:36.929Z',
    },
    coverImageUrl: 'https://i.imgur.com/bqgW6JXh.jpeg',
    title: 'Hello world 2',
    body: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut a',
    views: 12,
    likes: 3,
    likeId: null,
    createdAt: '2022-06-27T06:32:38.231Z',
    lastModifiedAt: '2022-06-27T06:32:38.231Z',
  },
]

export default articles
