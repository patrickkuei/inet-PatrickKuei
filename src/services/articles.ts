import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import { IArticleViewModel } from './types/articles/i-article.view-model'
import { IGetArticlesQuery } from './types/articles/i-get-articles.query'

export const articlesApi = createApi({
  reducerPath: 'articlesApi',
  baseQuery: fetchBaseQuery({ baseUrl: 'http://3.112.59.180:33823/' }),
  endpoints: (builder) => ({
    getArticles: builder.query<IArticleViewModel[], IGetArticlesQuery>({
      query: (params) => ({
        url: 'api/articles',
        params,
      }),
    }),
  }),
})

export const { useGetArticlesQuery } = articlesApi
