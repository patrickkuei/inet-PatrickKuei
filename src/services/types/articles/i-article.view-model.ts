import { IArticleCategoryViewModel } from '../shared/i-article-category.view-model'
import { IAuthorViewModel } from '../shared/i-author.view-model'

export interface IArticleViewModel {
  id: number
  category: IArticleCategoryViewModel
  author: IAuthorViewModel
  coverImageUrl: string
  title: string
  body: string
  views: number
  likes: number
  likeId: number | null
  createdAt: Date
  lastModifiedAt: string | null
}
