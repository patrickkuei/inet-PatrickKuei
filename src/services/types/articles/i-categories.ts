export interface ICategory {
  title: string
  imgSrc: string
}

export type ICategories = ICategory[]
