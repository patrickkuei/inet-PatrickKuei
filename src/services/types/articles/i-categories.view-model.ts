export interface ICategory {
  id: number
  code: string
  imageUrl: string
}

export type ICategoriesRespont = ICategory[]
