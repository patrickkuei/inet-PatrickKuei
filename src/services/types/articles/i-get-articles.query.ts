import { IPagableQuery } from '@inet/services/types/shared/pagination'

export enum ArticleCreatedWithin {
  Today = 'today',
  Week = 'week',
  Month = 'month',
  Year = 'year',
  All = 'all',
}

export interface IGetArticlesQuery extends IPagableQuery {
  createdWithin?: ArticleCreatedWithin
  categoryId?: number
  keyword?: string
  authorUsername?: string
}
