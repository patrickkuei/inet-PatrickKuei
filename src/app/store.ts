import articleReducer from '@inet/redux/slices/articleSlice'
import paginationReducer from '@inet/redux/slices/paginationSlice'
import { apiSlice } from '@inet/services/apiSlice'
import { configureStore } from '@reduxjs/toolkit'

export const store = configureStore({
  reducer: {
    [apiSlice.reducerPath]: apiSlice.reducer,
    articleReducer,
    paginationReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(apiSlice.middleware),
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
