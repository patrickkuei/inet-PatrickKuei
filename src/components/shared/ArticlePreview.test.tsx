import { store } from '@inet/app/store'
import articles from '@inet/mockData/articles'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import ArticlePreview from './ArticlePreview'

it('render ArticlePreview', () => {
  const component = renderer.create(
    <Provider store={store}>
      <ArticlePreview article={articles[0]} createdAt="8 days ago" />
    </Provider>,
  )

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
