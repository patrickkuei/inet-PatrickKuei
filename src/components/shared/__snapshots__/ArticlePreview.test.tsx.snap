// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ArticlePreview 1`] = `
<div
  className="text-left h-24 flex flex-row gap-x-2"
>
  <div
    className="w-24 shrink-0 bg-cyan-50 flex items-center justify-center"
  >
    <img
      alt="Hello world"
      className="max-h-full"
      src="https://i.imgur.com/bqgW6JXh.jpeg"
    />
  </div>
  <div
    className="flex flex-col grow gap-y-1 min-w-0"
  >
    <div
      className="flex items-center gap-x-1 text-xs"
    >
      <div
        className="flex items-center gap-x-1 shrink-0"
      >
        <img
          alt="animal"
          className="w-4 h-4 border border-gray-100 rounded inline"
          src="https://i.imgur.com/JR2xyV6s.jpg"
        />
        <span>
          animal
        </span>
      </div>
      <span
        className="text-gray-400"
      >
        •
      </span>
      <div
        className="flex items-center gap-x-1 min-w-0"
      >
        <img
          alt="人類一號"
          className="w-4 h-4 border border-gray-100 rounded-full inline"
          src="https://i.imgur.com/1yJEtrqs.jpeg"
        />
        <span
          className="truncate"
        >
          人類一號
        </span>
      </div>
      <span
        className="text-gray-400"
      >
        •
      </span>
      <span
        className="whitespace-nowrap"
      >
        8 days ago
      </span>
    </div>
    <div>
      <div
        className="w-fit text-tiny font-bold line-clamp-2 hover:text-primary-400 hover:cursor-pointer"
      >
        Hello world
      </div>
      <div
        className="text-xs text-gray-300 truncate"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut a
      </div>
    </div>
    <div
      className="text-mini flex grow items-end gap-x-4"
    >
      <span
        className="flex gap-x-1"
      >
        <svg
          data-jest-file-name="eye.svg"
          data-jest-svg-name="eye"
          data-testid="eye"
        />
        123
      </span>
      <span
        className="flex gap-x-1"
      >
        <svg
          data-jest-file-name="heart.svg"
          data-jest-svg-name="heart"
          data-testid="heart"
        />
        45
      </span>
    </div>
  </div>
</div>
`;
